File: backend/app/routers/devices.py
import csv
import io
from fastapi import APIRouter, Depends, HTTPException, Query, Request, UploadFile, File
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.database import get_db
from app.models import Device, Backup, User, Location
from app.schemas import DeviceCreate, DeviceUpdate, DeviceOut, ScheduleUpdate
from app.auth import get_current_user
from app.services.backup_service import run_backup
from app.services.fortigate import test_fortigate
from app.services.juniper import test_juniper
from app.services.cisco import test_cisco
from app.services.paloalto import test_paloalto
from app.services.audit_service import log_action
from app.crypto import encrypt, decrypt

ENCRYPTED_FIELDS = ("auth_token", "ssh_password", "enable_password")


class DecryptedDevice:
    def __init__(self, device):
        self._device = device
    def __getattr__(self, name):
        val = getattr(self._device, name)
        if name in ENCRYPTED_FIELDS and val:
            return decrypt(val)
        return val

router = APIRouter(prefix="/api/devices", tags=["devices"])


def enrich_device(device: Device, db: Session) -> dict:
    last = db.query(Backup.created_at).filter(Backup.device_id == device.id, Backup.status == "success").order_by(Backup.created_at.desc()).first()
    success_count = db.query(func.count(Backup.id)).filter(Backup.device_id == device.id, Backup.status == "success").scalar()
    failed_count = db.query(func.count(Backup.id)).filter(Backup.device_id == device.id, Backup.status == "failed").scalar()
    d = {c.name: getattr(device, c.name) for c in device.__table__.columns}
    d["last_backup"] = last[0] if last else None
    d["backup_count"] = success_count or 0
    d["failed_count"] = failed_count or 0
    d["location_name"] = device.location.name if device.location else None
    d["has_token"] = bool(device.auth_token)
    d["has_ssh_password"] = bool(device.ssh_password)
    d["has_enable_password"] = bool(device.enable_password)
    return d


@router.get("", response_model=list[DeviceOut])
def list_devices(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    devices = db.query(Device).order_by(Device.name).all()
    return [enrich_device(d, db) for d in devices]


@router.post("", response_model=DeviceOut, status_code=201)
def create_device(body: DeviceCreate, request: Request, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if db.query(Device).filter(Device.name == body.name).first():
        raise HTTPException(status_code=400, detail="Device name already exists")
    data = body.model_dump()
    for field in ENCRYPTED_FIELDS:
        if data.get(field):
            data[field] = encrypt(data[field])
    device = Device(**data)
    db.add(device)
    db.commit()
    db.refresh(device)
    log_action(db, user, "create", "device", device.name, f"Vendor: {device.vendor}, IP: {device.ip_address}", request.client.host)
    return enrich_device(device, db)


@router.get("/{device_id}", response_model=DeviceOut)
def get_device(device_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    return enrich_device(device, db)


@router.put("/{device_id}", response_model=DeviceOut)
def update_device(device_id: int, body: DeviceUpdate, request: Request, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    changes = body.model_dump(exclude_unset=True)
    changed_fields = [k for k in changes if k not in ENCRYPTED_FIELDS]
    for field, value in changes.items():
        if field in ENCRYPTED_FIELDS and value:
            value = encrypt(value)
        setattr(device, field, value)
    db.commit()
    db.refresh(device)
    log_action(db, user, "update", "device", device.name, f"Değişen: {', '.join(changed_fields) or 'credentials'}", request.client.host)
    return enrich_device(device, db)


@router.delete("/{device_id}")
def delete_device(device_id: int, keep_backups: bool = Query(True), request: Request = None, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    device_name = device.name
    if not keep_backups:
        import shutil
        from app.config import BACKUP_DIR
        device_dir = BACKUP_DIR / device.vendor / device.name
        if device_dir.exists():
            shutil.rmtree(device_dir)
    db.delete(device)
    db.commit()
    log_action(db, user, "delete", "device", device_name, f"Yedekler korundu: {'Evet' if keep_backups else 'Hayır'}", request.client.host if request else None)
    return {"message": "Device deleted"}


@router.post("/{device_id}/backup")
def trigger_backup(device_id: int, request: Request = None, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    result = run_backup(DecryptedDevice(device), db, triggered_by="manual")
    log_action(db, user, "backup", "device", device.name, f"Sonuç: {result['status']}", request.client.host if request else None)
    return result


@router.post("/{device_id}/test")
def test_connection(device_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    try:
        dd = DecryptedDevice(device)
        if device.vendor == "fortigate":
            test_fortigate(dd)
        elif device.vendor == "juniper":
            test_juniper(dd)
        elif device.vendor == "cisco":
            test_cisco(dd)
        elif device.vendor == "paloalto":
            test_paloalto(dd)
        return {"status": "success", "message": "Connection successful"}
    except Exception as e:
        return {"status": "failed", "message": str(e)}


@router.put("/{device_id}/schedule")
def set_schedule(device_id: int, body: ScheduleUpdate, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    device.schedule_cron = body.cron
    db.commit()
    from app.services.scheduler_service import reschedule_device
    reschedule_device(device)
    return {"message": "Schedule updated", "cron": body.cron}


@router.delete("/{device_id}/schedule")
def remove_schedule(device_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    device = db.query(Device).filter(Device.id == device_id).first()
    if not device:
        raise HTTPException(status_code=404, detail="Device not found")
    device.schedule_cron = None
    db.commit()
    from app.services.scheduler_service import remove_device_schedule
    remove_device_schedule(device.id)
    return {"message": "Schedule removed"}


CSV_COLUMNS = ["name", "vendor", "ip_address", "port", "platform", "auth_token",
               "ssh_username", "ssh_password", "enable_password", "location", "vdom"]
VALID_VENDORS = {"fortigate", "juniper", "cisco", "paloalto"}
VALID_PLATFORMS = {"ios", "nxos", "asa"}


@router.get("/bulk/template")
def download_csv_template(user: User = Depends(get_current_user)):
    buf = io.StringIO()
    w = csv.writer(buf)
    w.writerow(CSV_COLUMNS)
    w.writerow(["FW-Istanbul", "fortigate", "10.0.1.1", "443", "", "api-key-here", "", "", "", "Istanbul DC", ""])
    w.writerow(["SW-Core", "cisco", "10.0.1.2", "22", "ios", "", "admin", "pass123", "enable123", "Ankara DC", ""])
    w.writerow(["JUN-Edge", "juniper", "10.0.1.3", "22", "", "", "admin", "pass456", "", "", ""])
    buf.seek(0)
    return StreamingResponse(
        iter([buf.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=confbox_devices_template.csv"},
    )


@router.post("/bulk/preview")
async def bulk_preview(file: UploadFile = File(...), db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if not file.filename.endswith(".csv"):
        raise HTTPException(status_code=400, detail="Sadece CSV dosyası yüklenebilir")
    content = (await file.read()).decode("utf-8-sig", errors="replace")
    reader = csv.DictReader(io.StringIO(content))
    existing_names = {d.name for d in db.query(Device.name).all()}
    location_map = {loc.name.lower(): loc.id for loc in db.query(Location).all()}
    rows = []
    for i, row in enumerate(reader):
        errors = []
        name = row.get("name", "").strip()
        vendor = row.get("vendor", "").strip().lower()
        ip = row.get("ip_address", "").strip()
        port_str = row.get("port", "").strip()
        platform = row.get("platform", "").strip().lower() or "ios"
        if not name:
            errors.append("Ad boş")
        elif name in existing_names:
            errors.append("Bu isimde cihaz var")
        if vendor not in VALID_VENDORS:
            errors.append(f"Geçersiz vendor: {vendor}")
        if not ip:
            errors.append("IP boş")
        port = 0
        try:
            port = int(port_str) if port_str else (443 if vendor == "fortigate" else 22)
        except ValueError:
            errors.append("Geçersiz port")
        if vendor == "cisco" and platform not in VALID_PLATFORMS:
            errors.append(f"Geçersiz platform: {platform}")
        loc_name = row.get("location", "").strip()
        loc_id = location_map.get(loc_name.lower()) if loc_name else None
        rows.append({
            "row": i + 1,
            "name": name, "vendor": vendor, "ip_address": ip, "port": port,
            "platform": platform,
            "auth_token": bool(row.get("auth_token", "").strip()),
            "ssh_username": row.get("ssh_username", "").strip(),
            "ssh_password": bool(row.get("ssh_password", "").strip()),
            "enable_password": bool(row.get("enable_password", "").strip()),
            "location": loc_name, "location_id": loc_id,
            "vdom": row.get("vdom", "").strip(),
            "errors": errors,
            "valid": len(errors) == 0,
        })
        existing_names.add(name)
    return {"rows": rows, "total": len(rows), "valid": sum(1 for r in rows if r["valid"]), "invalid": sum(1 for r in rows if not r["valid"])}


@router.post("/bulk/import")
async def bulk_import(file: UploadFile = File(...), request: Request = None, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    if not file.filename.endswith(".csv"):
        raise HTTPException(status_code=400, detail="Sadece CSV dosyası yüklenebilir")
    content = (await file.read()).decode("utf-8-sig", errors="replace")
    reader = csv.DictReader(io.StringIO(content))
    existing_names = {d.name for d in db.query(Device.name).all()}
    location_map = {loc.name.lower(): loc.id for loc in db.query(Location).all()}
    created, skipped = 0, 0
    for row in reader:
        name = row.get("name", "").strip()
        vendor = row.get("vendor", "").strip().lower()
        ip = row.get("ip_address", "").strip()
        port_str = row.get("port", "").strip()
        platform = row.get("platform", "").strip().lower() or "ios"
        if not name or not ip or vendor not in VALID_VENDORS or name in existing_names:
            skipped += 1
            continue
        try:
            port = int(port_str) if port_str else (443 if vendor == "fortigate" else 22)
        except ValueError:
            skipped += 1
            continue
        loc_name = row.get("location", "").strip()
        loc_id = location_map.get(loc_name.lower()) if loc_name else None
        data = {
            "name": name, "vendor": vendor, "ip_address": ip, "port": port,
            "platform": platform, "location_id": loc_id,
            "vdom": row.get("vdom", "").strip() or None,
            "auth_token": row.get("auth_token", "").strip() or None,
            "ssh_username": row.get("ssh_username", "").strip() or None,
            "ssh_password": row.get("ssh_password", "").strip() or None,
            "enable_password": row.get("enable_password", "").strip() or None,
        }
        for field in ENCRYPTED_FIELDS:
            if data.get(field):
                data[field] = encrypt(data[field])
        device = Device(**data)
        db.add(device)
        existing_names.add(name)
        created += 1
    db.commit()
    log_action(db, user, "bulk_import", "device", f"{created} cihaz", f"Eklenen: {created}, Atlanan: {skipped}", request.client.host if request else None)
    return {"created": created, "skipped": skipped}

File: backend/internal/handler/devices.go
package handler

import (
	"encoding/csv"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"

	"github.com/yunuskargi/confbox/internal/auth"
	"github.com/yunuskargi/confbox/internal/crypto"
	"github.com/yunuskargi/confbox/internal/database"
	"github.com/yunuskargi/confbox/internal/models"
	"github.com/yunuskargi/confbox/internal/service"
)

var encryptedFields = []string{"auth_token", "ssh_password", "enable_password"}

func enrichDevice(id int) models.DeviceOut {
	var out models.DeviceOut
	row := database.DB.QueryRow(`
		SELECT d.id, d.name, d.vendor, d.ip_address, d.port, d.location_id, l.name,
		       d.vdom, d.platform, d.schedule_cron, d.is_active, d.created_at, d.updated_at,
		       d.auth_token, d.ssh_password, d.enable_password
		FROM devices d LEFT JOIN locations l ON l.id = d.location_id
		WHERE d.id = ?`, id)

	var locID *int
	var locName, vdom, platform, cron *string
	var authToken, sshPass, enablePass *string
	row.Scan(&out.ID, &out.Name, &out.Vendor, &out.IPAddress, &out.Port,
		&locID, &locName, &vdom, &platform, &cron, &out.IsActive, &out.CreatedAt, &out.UpdatedAt,
		&authToken, &sshPass, &enablePass)
	out.LocationID = locID
	out.LocationName = locName
	out.Vdom = vdom
	out.Platform = platform
	out.ScheduleCron = cron
	out.HasToken = authToken != nil && *authToken != ""
	out.HasSSHPassword = sshPass != nil && *sshPass != ""
	out.HasEnablePassword = enablePass != nil && *enablePass != ""

	database.DB.Get(&out.BackupCount, "SELECT COUNT(*) FROM backups WHERE device_id = ? AND status = 'success'", id)
	database.DB.Get(&out.FailedCount, "SELECT COUNT(*) FROM backups WHERE device_id = ? AND status = 'failed'", id)

	var lastBackup *string
	database.DB.Get(&lastBackup, "SELECT created_at FROM backups WHERE device_id = ? AND status = 'success' ORDER BY created_at DESC LIMIT 1", id)
	out.LastBackup = lastBackup

	return out
}

func ListDevices(w http.ResponseWriter, r *http.Request) {
	var ids []int
	database.DB.Select(&ids, "SELECT id FROM devices ORDER BY name")
	result := make([]models.DeviceOut, len(ids))
	for i, id := range ids {
		result[i] = enrichDevice(id)
	}
	writeJSON(w, 200, result)
}

type deviceCreateBody struct {
	Name           string  `json:"name"`
	Vendor         string  `json:"vendor"`
	IPAddress      string  `json:"ip_address"`
	Port           int     `json:"port"`
	LocationID     *int    `json:"location_id"`
	Vdom           *string `json:"vdom"`
	AuthToken      *string `json:"auth_token"`
	SSHUsername    *string `json:"ssh_username"`
	SSHPassword    *string `json:"ssh_password"`
	EnablePassword *string `json:"enable_password"`
	Platform       *string `json:"platform"`
	ScheduleCron   *string `json:"schedule_cron"`
}

func CreateDevice(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	var body deviceCreateBody
	if err := decodeBody(r, &body); err != nil {
		writeError(w, 400, "Invalid request body")
		return
	}

	if strings.ContainsAny(body.Name, "/\\..") || body.Name == "" {
		writeError(w, 400, "Invalid device name")
		return
	}

	validVendorsSet := map[string]bool{"fortigate": true, "juniper": true, "cisco": true, "paloalto": true}
	if !validVendorsSet[body.Vendor] {
		writeError(w, 400, "Invalid vendor")
		return
	}

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE name = ?", body.Name)
	if exists > 0 {
		writeError(w, 400, "Device name already exists")
		return
	}

	if body.Port == 0 {
		body.Port = 443
	}
	platform := "ios"
	if body.Platform != nil {
		platform = *body.Platform
	}

	authToken := encryptOptional(body.AuthToken)
	sshPass := encryptOptional(body.SSHPassword)
	enablePass := encryptOptional(body.EnablePassword)

	res, _ := database.DB.Exec(`INSERT INTO devices (name, vendor, ip_address, port, location_id, vdom, auth_token, ssh_username, ssh_password, enable_password, platform, schedule_cron, is_active, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, datetime('now'), datetime('now'))`,
		body.Name, body.Vendor, body.IPAddress, body.Port, body.LocationID, body.Vdom,
		authToken, body.SSHUsername, sshPass, enablePass, platform, body.ScheduleCron)

	id, _ := res.LastInsertId()
	uid := user.ID
	service.LogAction(&uid, user.Username, "create", "device", body.Name,
		fmt.Sprintf("Vendor: %s, IP: %s", body.Vendor, body.IPAddress), clientIP(r))

	if body.ScheduleCron != nil && *body.ScheduleCron != "" {
		service.ScheduleDevice(int(id), *body.ScheduleCron)
	}

	writeJSON(w, 201, enrichDevice(int(id)))
}

func GetDevice(w http.ResponseWriter, r *http.Request) {
	id := paramInt(r, "id")
	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}
	writeJSON(w, 200, enrichDevice(id))
}

var allowedDeviceColumns = map[string]bool{
	"name": true, "ip_address": true, "port": true, "location_id": true,
	"vdom": true, "auth_token": true, "ssh_username": true, "ssh_password": true,
	"enable_password": true, "platform": true, "schedule_cron": true, "is_active": true,
}

func UpdateDevice(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	id := paramInt(r, "id")

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}

	var body map[string]any
	if err := decodeBody(r, &body); err != nil {
		writeError(w, 400, "Invalid request body")
		return
	}

	for key, val := range body {
		if val == nil {
			continue
		}
		if !allowedDeviceColumns[key] {
			continue
		}
		switch key {
		case "auth_token", "ssh_password", "enable_password":
			if s, ok := val.(string); ok && s != "" {
				val = crypto.Encrypt(s)
			}
		case "port":
			if f, ok := val.(float64); ok {
				val = int(f)
			}
		case "location_id":
			if f, ok := val.(float64); ok {
				val = int(f)
			}
		case "is_active":
			if b, ok := val.(bool); ok {
				if b {
					val = 1
				} else {
					val = 0
				}
			}
		}
		database.DB.Exec(fmt.Sprintf("UPDATE devices SET %s = ? WHERE id = ?", key), val, id)
	}
	database.DB.Exec("UPDATE devices SET updated_at = datetime('now') WHERE id = ?", id)

	var name string
	database.DB.Get(&name, "SELECT name FROM devices WHERE id = ?", id)
	uid := user.ID
	service.LogAction(&uid, user.Username, "update", "device", name, "Device updated", clientIP(r))

	if cronVal, ok := body["schedule_cron"]; ok {
		if cronStr, ok := cronVal.(string); ok && cronStr != "" {
			service.ScheduleDevice(id, cronStr)
		} else {
			service.RemoveDeviceSchedule(id)
		}
	}

	writeJSON(w, 200, enrichDevice(id))
}

func DeleteDevice(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	id := paramInt(r, "id")
	keepBackups := r.URL.Query().Get("keep_backups") != "false"

	var device struct {
		Name   string `db:"name"`
		Vendor string `db:"vendor"`
	}
	err := database.DB.Get(&device, "SELECT name, vendor FROM devices WHERE id = ?", id)
	if err != nil {
		writeError(w, 404, "Device not found")
		return
	}

	if !keepBackups {
		service.DeleteDeviceBackupFiles(device.Vendor, device.Name)
	}

	service.RemoveDeviceSchedule(id)
	database.DB.Exec("DELETE FROM backups WHERE device_id = ?", id)
	database.DB.Exec("DELETE FROM devices WHERE id = ?", id)

	uid := user.ID
	service.LogAction(&uid, user.Username, "delete", "device", device.Name,
		fmt.Sprintf("Backups kept: %v", keepBackups), clientIP(r))
	writeJSON(w, 200, map[string]string{"message": "Device deleted"})
}

func TriggerBackup(w http.ResponseWriter, r *http.Request) {
	user := auth.GetUser(r)
	id := paramInt(r, "id")

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}

	result := service.RunBackup(id, "manual")

	var name string
	database.DB.Get(&name, "SELECT name FROM devices WHERE id = ?", id)
	uid := user.ID
	service.LogAction(&uid, user.Username, "backup", "device", name,
		fmt.Sprintf("Result: %s", result["status"]), clientIP(r))

	writeJSON(w, 200, result)
}

func TestConnection(w http.ResponseWriter, r *http.Request) {
	id := paramInt(r, "id")

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}

	result := service.TestDeviceConnection(id)
	writeJSON(w, 200, result)
}

type scheduleBody struct {
	Cron string `json:"cron"`
}

func SetSchedule(w http.ResponseWriter, r *http.Request) {
	id := paramInt(r, "id")
	var body scheduleBody
	if err := decodeBody(r, &body); err != nil {
		writeError(w, 400, "Invalid request body")
		return
	}

	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}

	database.DB.Exec("UPDATE devices SET schedule_cron = ?, updated_at = datetime('now') WHERE id = ?", body.Cron, id)
	service.ScheduleDevice(id, body.Cron)
	writeJSON(w, 200, map[string]string{"message": "Schedule updated", "cron": body.Cron})
}

func RemoveSchedule(w http.ResponseWriter, r *http.Request) {
	id := paramInt(r, "id")
	var exists int
	database.DB.Get(&exists, "SELECT COUNT(*) FROM devices WHERE id = ?", id)
	if exists == 0 {
		writeError(w, 404, "Device not found")
		return
	}

	database.DB.Exec("UPDATE devices SET schedule_cron = NULL, updated_at = datetime('now') WHERE id = ?", id)
	service.RemoveDeviceSchedule(id)
	writeJSON(w, 200, map[string]string{"message": "Schedule removed"})
}

func BulkTemplate(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/csv")
	w.Header().Set("Content-Disposition", "attachment; filename=confbox_devices_template.csv")
	wr := csv.NewWriter(w)
	wr.Write([]string{"name", "vendor", "ip_address", "port", "platform", "auth_token", "ssh_username", "ssh_password", "enable_password", "location", "vdom"})
	wr.Write([]string{"FW-Istanbul", "fortigate", "10.0.1.1", "443", "", "api-key-here", "", "", "", "Istanbul DC", ""})
	wr.Write([]string{"SW-Core", "cisco", "10.0.1.2", "22", "ios", "", "admin", "pass123", "enable123", "Ankara DC", ""})
	wr.Write([]string{"JUN-Edge", "juniper", "10.0.1.3", "22", "", "", "admin", "pass456", "", "", ""})
	wr.Flush()
}

func BulkPreview(w http.ResponseWriter, r *http.Request) {
	r.Body = http.MaxBytesReader(w, r.Body, 5<<20)
	file, header, err := r.FormFile("file")
	if err != nil {
		writeError(w, 400, "No file uploaded")
		return
	}
	defer file.Close()

	if !strings.HasSuffix(header.Filename, ".csv") {
		writeError(w, 400, "Only CSV files accepted")
		return
	}

	content, _ := io.ReadAll(io.LimitReader(file, 5<<20))
	reader := csv.NewReader(strings.NewReader(string(content)))
	headers, _ := reader.Read()
	colMap := make(map[string]int)
	for i, h := range headers {
		colMap[strings.TrimSpace(strings.ToLower(h))] = i
	}

	var existingNames []string
	database.DB.Select(&existingNames, "SELECT name FROM devices")
	nameSet := make(map[string]bool)
	for _, n := range existingNames {
		nameSet[n] = true
	}

	type locRow struct {
		ID   int    `db:"id"`
		Name string `db:"name"`
	}
	var locs []locRow
	database.DB.Select(&locs, "SELECT id, name FROM locations")
	locMap := make(map[string]int)
	for _, l := range locs {
		locMap[strings.ToLower(l.Name)] = l.ID
	}

	validVendors := map[string]bool{"fortigate": true, "juniper": true, "cisco": true, "paloalto": true}
	validPlatforms := map[string]bool{"ios": true, "nxos": true, "asa": true}

	var rows []map[string]any
	rowNum := 0
	for {
		record, err := reader.Read()
		if err != nil {
			break
		}
		rowNum++
		getCol := func(name string) string {
			if idx, ok := colMap[name]; ok && idx < len(record) {
				return strings.TrimSpace(record[idx])
			}
			return ""
		}

		name := getCol("name")
		vendor := strings.ToLower(getCol("vendor"))
		ip := getCol("ip_address")
		portStr := getCol("port")
		platform := strings.ToLower(getCol("platform"))
		if platform == "" {
			platform = "ios"
		}

		var errors []string
		if name == "" {
			errors = append(errors, "Name empty")
		} else if nameSet[name] {
			errors = append(errors, "Device name exists")
		}
		if !validVendors[vendor] {
			errors = append(errors, "Invalid vendor: "+vendor)
		}
		if ip == "" {
			errors = append(errors, "IP empty")
		}
		port := 0
		if portStr != "" {
			p, err := strconv.Atoi(portStr)
			if err != nil {
				errors = append(errors, "Invalid port")
			} else {
				port = p
			}
		} else if vendor == "fortigate" {
			port = 443
		} else {
			port = 22
		}
		if vendor == "cisco" && !validPlatforms[platform] {
			errors = append(errors, "Invalid platform: "+platform)
		}

		locName := getCol("location")
		var locID *int
		if locName != "" {
			if lid, ok := locMap[strings.ToLower(locName)]; ok {
				locID = &lid
			}
		}

		row := map[string]any{
			"row": rowNum, "name": name, "vendor": vendor, "ip_address": ip, "port": port,
			"platform": platform, "auth_token": getCol("auth_token") != "",
			"ssh_username": getCol("ssh_username"), "ssh_password": getCol("ssh_password") != "",
			"enable_password": getCol("enable_password") != "",
			"location": locName, "location_id": locID, "vdom": getCol("vdom"),
			"errors": errors, "valid": len(errors) == 0,
		}
		rows = append(rows, row)
		nameSet[name] = true
	}

	valid := 0
	for _, r := range rows {
		if r["valid"].(bool) {
			valid++
		}
	}
	writeJSON(w, 200, map[string]any{"rows": rows, "total": len(rows), "valid": valid, "invalid": len(rows) - valid})
}

func BulkImport(w http.ResponseWriter, r *http.Request) {
	r.Body = http.MaxBytesReader(w, r.Body, 5<<20) // 5MB limit
	user := auth.GetUser(r)
	file, header, err := r.FormFile("file")
	if err != nil {
		writeError(w, 400, "No file uploaded")
		return
	}
	defer file.Close()

	if !strings.HasSuffix(header.Filename, ".csv") {
		writeError(w, 400, "Only CSV files accepted")
		return
	}

	content, _ := io.ReadAll(io.LimitReader(file, 5<<20))
	reader := csv.NewReader(strings.NewReader(string(content)))
	headers, _ := reader.Read()
	colMap := make(map[string]int)
	for i, h := range headers {
		colMap[strings.TrimSpace(strings.ToLower(h))] = i
	}

	var existingNames []string
	database.DB.Select(&existingNames, "SELECT name FROM devices")
	nameSet := make(map[string]bool)
	for _, n := range existingNames {
		nameSet[n] = true
	}

	type locRow struct {
		ID   int    `db:"id"`
		Name string `db:"name"`
	}
	var locs []locRow
	database.DB.Select(&locs, "SELECT id, name FROM locations")
	locMap := make(map[string]int)
	for _, l := range locs {
		locMap[strings.ToLower(l.Name)] = l.ID
	}

	validVendors := map[string]bool{"fortigate": true, "juniper": true, "cisco": true, "paloalto": true}
	created, skipped := 0, 0

	for {
		record, err := reader.Read()
		if err != nil {
			break
		}
		getCol := func(name string) string {
			if idx, ok := colMap[name]; ok && idx < len(record) {
				return strings.TrimSpace(record[idx])
			}
			return ""
		}

		name := getCol("name")
		vendor := strings.ToLower(getCol("vendor"))
		ip := getCol("ip_address")
		portStr := getCol("port")
		platform := strings.ToLower(getCol("platform"))
		if platform == "" {
			platform = "ios"
		}

		if name == "" || ip == "" || !validVendors[vendor] || nameSet[name] {
			skipped++
			continue
		}

		port := 22
		if portStr != "" {
			p, err := strconv.Atoi(portStr)
			if err != nil {
				skipped++
				continue
			}
			port = p
		} else if vendor == "fortigate" {
			port = 443
		}

		locName := getCol("location")
		var locID *int
		if locName != "" {
			if lid, ok := locMap[strings.ToLower(locName)]; ok {
				locID = &lid
			}
		}

		authToken := encryptIfNotEmpty(getCol("auth_token"))
		sshPass := encryptIfNotEmpty(getCol("ssh_password"))
		enablePass := encryptIfNotEmpty(getCol("enable_password"))
		sshUser := getCol("ssh_username")
		vdom := getCol("vdom")

		database.DB.Exec(`INSERT INTO devices (name, vendor, ip_address, port, location_id, vdom, auth_token, ssh_username, ssh_password, enable_password, platform, is_active, created_at, updated_at)
			VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, datetime('now'), datetime('now'))`,
			name, vendor, ip, port, locID, nilIfEmpty(vdom), authToken, nilIfEmpty(sshUser), sshPass, enablePass, platform)

		nameSet[name] = true
		created++
	}

	uid := user.ID
	service.LogAction(&uid, user.Username, "bulk_import", "device",
		fmt.Sprintf("%d devices", created), fmt.Sprintf("Created: %d, Skipped: %d", created, skipped), clientIP(r))
	writeJSON(w, 200, map[string]any{"created": created, "skipped": skipped})
}

func encryptOptional(s *string) *string {
	if s == nil || *s == "" {
		return s
	}
	enc := crypto.Encrypt(*s)
	return &enc
}

func encryptIfNotEmpty(s string) *string {
	if s == "" {
		return nil
	}
	enc := crypto.Encrypt(s)
	return &enc
}

func nilIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}
